grpcgcp/gcp_interceptor_test.go

```diff
@@ -19,130 +19,229 @@
 	}
-	if affCfg.GetCommand() != configpb.AffinityConfig_BOUND {
-		t.Errorf("affinity config has incorrect command: %v, want: %v", affCfg.GetCommand(), configpb.AffinityConfig_BOUND)
+	if err := cs.RecvMsg(wantRes); err != nil {
+		t.Fatalf("RecvMsg() returned error: %v, want: nil", err)
 	}
+}
 
-	methodName = "method3"
-	affCfg, ok = gcpInt.methodToAffinity[methodName]
-	if !ok {
-		t.Fatalf("gcpInt.methodToAffinity should contain key: %v", methodName)
-	}
-	if affCfg.GetCommand() != configpb.AffinityConfig_UNBIND {
-		t.Errorf("affinity config has incorrect command: %v, want: %v", affCfg.GetCommand(), configpb.AffinityConfig_UNBIND)
-	}
+func TestGCPStreamClientInterceptorCallingReadBeforeSend(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	ctx := context.TODO()
+	wantMethod := "someMethod"
+	wantReq := "someRequest"
+	wantRes := &fakeResp{}
+	wantGcpCtx := &gcpContext{
+		reqMsg: wantReq,
+	}
+	wantSD := &grpc.StreamDesc{}
+	wantCC := &grpc.ClientConn{}
+	wantOpts := []grpc.CallOption{grpc.CallContentSubtype("someSubtype"), grpc.MaxCallRecvMsgSize(42)}
+
+	streamerCalled := false
+	streamer := func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+		streamerCalled = true
+		gotGcpCtx, hasGcpCtx := ctx.Value(gcpKey).(*gcpContext)
+		if !hasGcpCtx {
+			t.Errorf("grpc.Streamer called with context without gcpContext")
+		} else if diff := cmp.Diff(wantGcpCtx, gotGcpCtx, cmp.AllowUnexported(gcpContext{})); diff != "" {
+			t.Errorf("grpc.Streamer called with unexpected gcpContext (-want, +got):\n%s", diff)
+		}
+		if desc != wantSD {
+			t.Errorf("grpc.Streamer called with unexpected StreamDesc: %v, want: %v", desc, wantSD)
+		}
+		if cc != wantCC {
+			t.Errorf("grpc.Streamer called with unexpected ClientConn: %v, want: %v", cc, wantCC)
+		}
+		if method != wantMethod {
+			t.Errorf("grpc.Streamer called with unexpected method: %v, want: %v", method, wantMethod)
+		}
+		if diff := cmp.Diff(wantOpts, opts); diff != "" {
+			t.Errorf("grpc.Streamer called with unexpected options (-want, +got):\n%s", diff)
+		}
+		mockCS := mocks.NewMockClientStream(mockCtrl)
+		mockCS.EXPECT().SendMsg(gomock.Eq(wantReq)).Times(1)
+		mockCS.EXPECT().RecvMsg(&strictMatcher{matchWith: wantRes}).Times(1)
+		return mockCS, nil
+	}
+	cs, err := GCPStreamClientInterceptor(
+		ctx,
+		wantSD,
+		wantCC,
+		wantMethod,
+		streamer,
+		wantOpts...,
+	)
+	if err != nil {
+		t.Fatalf("GCPStreamClientInterceptor(...) returned error: %v, want: nil", err)
+	}
+	if streamerCalled {
+		t.Fatalf("GCPStreamClientInterceptor(...) unexpectedly called grpc.Streamer on init")
+	}
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	received := &sync.WaitGroup{}
+	received.Add(1)
+	go func() {
+		wg.Done()
+		if err := cs.RecvMsg(wantRes); err != nil {
+			t.Errorf("RecvMsg() returned error: %v, want: nil", err)
+		}
+		received.Done()
+	}()
+	wg.Wait()
```

Comment:
I want to make sure `RecvMsg` is called before `SendMsg` because `RecvMsg` must block until `SendMsg` is called and this test should validate that it will unblock after `SendMsg`.

I don't think it will cause flakiness, but it does not guarantee `RecvMsg` will be called before `SendMsg`, so adding a 1ms delay as well.